kolibri/core/content/utils/content_manifest.py

```diff
@@ -0,0 +1,195 @@
+import hashlib
+import itertools
+import json
+import os
+from collections import namedtuple
+
+from kolibri.core.content.models import ChannelMetadata
+from kolibri.core.content.models import ContentNode
+
+try:
+    FileNotFoundError
+except NameError:
+    FileNotFoundError = IOError
+
+
+ContentManifestChannelData = namedtuple(
+    "ContentManifestChannelData", ["channel_id", "channel_version", "include_node_ids"]
+)
+
+
+class ContentManifestParseError(ValueError):
+    pass
+
+
+class ContentManifest(object):
+    """
+    A content manifest describes a selection of Kolibri content across
+    multiple channels. It is usually stored as a JSON file along with exported
+    content. It contains a list of channels, with their versions, and a list
+    of node IDs associated with each channel ID + version.
+
+    When adding content to the manifest, this implementation optimizes the list
+    of node IDs to contain the smallest possible number of topic nodes, instead
+    of every included node. When nodes are added to a channel ID and version
+    that is already in the content manifest, those nodes will be added to the
+    existing list.
+    """
+
+    def __init__(self):
+        self._channels_dict = {}
+
+    def read(self, filenames):
+        if not isinstance(filenames, list):
+            filenames = [filenames]
+
+        files_read = []
+
+        for filename in filenames:
+            try:
+                with open(filename, "r") as fp:
+                    self.read_file(fp)
+            except FileNotFoundError:
+                pass
+            else:
+                files_read.append(filename)
+
+        return files_read
+
+    def read_file(self, fp):
+        json_str = fp.read()
+        if not json_str:
+            return
+        # Raises JSONDecodeError if the file is invalid
+        manifest_data = json.loads(json_str)
```

Comment:
Any reason not to just do a try...except here, catch any errors and just do `json.load` directly from the `fp` object?

Reply:
Yeah, that's funny it ended up looking like this :b Doesn't seem to be any reason, so I switched it to use `json.load(fp)`.